ksonnet-gen/ksonnet/emit.go

```diff
@@ -0,0 +1,588 @@
+) (*apiObject, error) {
+	if parsedName.Version == nil {
+		return nil, fmt.Errorf(
+			"Can't make API object from name with nil version. package type: '%d' kind: '%s'",
+			parsedName.PackageType,
+			parsedName.Kind)
+	}
+
+	var groupName kubespec.GroupName
+	if parsedName.Group == nil {
+		groupName = "core"
+	} else {
+		groupName = *parsedName.Group
+	}
+
+	group, ok := root.groups[groupName]
+	if !ok {
+		group = newGroup(groupName, root)
+		root.groups[groupName] = group
+	}
+
+	versionedAPI, ok := group.versionedAPIs[*parsedName.Version]
+	if !ok {
+		versionedAPI = newVersionedAPI(*parsedName.Version, group)
+		group.versionedAPIs[*parsedName.Version] = versionedAPI
+	}
+
+	apiObject, ok := versionedAPI.apiObjects[parsedName.Kind]
+	if ok {
+		log.Fatalf("Duplicate object kinds with name '%s'", parsedName.Unparse())
+	}
+	apiObject = newAPIObject(parsedName.Kind, versionedAPI, isTopLevel)
+	versionedAPI.apiObjects[parsedName.Kind] = apiObject
+	return apiObject, nil
+}
+
+func (root *root) getAPIObject(
+	parsedName *kubespec.ParsedDefinitionName,
+) (*apiObject, error) {
+	if parsedName.Version == nil {
+		return nil, fmt.Errorf(
+			"Can't make API object from name with nil version. package type: '%d' kind: '%s'",
+			parsedName.PackageType,
+			parsedName.Kind)
+	}
+
+	var groupName kubespec.GroupName
+	if parsedName.Group == nil {
+		groupName = "core"
+	} else {
+		groupName = *parsedName.Group
+	}
+
+	group, ok := root.groups[groupName]
+	if !ok {
+		return nil, fmt.Errorf(
+			"Could not retrieve object, group '%s' doesn't exist", groupName)
+	}
+
+	versionedAPI, ok := group.versionedAPIs[*parsedName.Version]
+	if !ok {
+		return nil, fmt.Errorf(
+			"Could not retrieve object, versioned API '%s' doesn't exist",
+			*parsedName.Version)
+	}
+
+	if apiObject, ok := versionedAPI.apiObjects[parsedName.Kind]; ok {
+		return apiObject, nil
+	}
+	return nil, fmt.Errorf(
+		"Could not retrieve object, kind '%s' doesn't exist",
+		parsedName.Kind)
+}
+
+//-----------------------------------------------------------------------------
+// Group.
+//-----------------------------------------------------------------------------
+
+// `group` is an abstract representation of a Kubernetes API group
+// (e.g., apps, extensions, core), which can be emitted as Jsonnet
+// code using the `emit` method.
+//
+// `group` contains a set of versioned APIs (e.g., v1, v1beta1, etc.),
+// though the logic for creating them is handled largely by `root`.
+type group struct {
+	name          kubespec.GroupName // e.g., core, apps, extensions.
+	versionedAPIs versionedAPISet    // e.g., v1, v1beta1.
+	parent        *root
+}
+type groupSet map[kubespec.GroupName]*group
+type groupSlice []*group
+
+func newGroup(name kubespec.GroupName, parent *root) *group {
+	return &group{
+		name:          name,
+		versionedAPIs: make(versionedAPISet),
+		parent:        parent,
+	}
+}
+
+func (group *group) emit(m *marshaller, root *root) error {
```

Comment:
Especially since only the `propertyMethod` needs it. For now I think the expedient thing to do is to just add a pointer to the `root` from the `propertyMethod`.